modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/planner/IndexRebuildPlannerTest.java

```diff
@@ -0,0 +1,367 @@
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteRel;
+import org.apache.ignite.internal.processors.query.calcite.rel.IgniteTableScan;
+import org.apache.ignite.internal.processors.query.calcite.schema.IgniteSchema;
+import org.apache.ignite.internal.processors.query.calcite.trait.IgniteDistribution;
+import org.apache.ignite.internal.processors.query.calcite.trait.IgniteDistributions;
+import org.apache.ignite.internal.processors.query.calcite.trait.TraitUtils;
+import org.apache.ignite.internal.processors.query.calcite.util.RexUtils;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.junit.Test;
+
+/**
+ * Planner test for index rebuild.
+ */
+public class IndexRebuildPlannerTest extends AbstractPlannerTest {
+    /** */
+    private IgniteSchema publicSchema;
+
+    /** */
+    private IndexRebuildAwareTable tbl;
+
+    /** {@inheritDoc} */
+    @Override public void setup() {
+        super.setup();
+
+        RelDataTypeFactory.Builder b = new RelDataTypeFactory.Builder(TYPE_FACTORY);
+
+        b.add("_KEY", TYPE_FACTORY.createJavaType(Object.class));
+        b.add("_VAL", TYPE_FACTORY.createJavaType(Object.class));
+        b.add("ID", TYPE_FACTORY.createSqlType(SqlTypeName.INTEGER));
+        b.add("VAL", TYPE_FACTORY.createSqlType(SqlTypeName.VARCHAR));
+
+        tbl = new IndexRebuildAwareTable("TBL", b.build(), 100d);
+
+        tbl.addIndex(RelCollations.of(ImmutableIntList.of(2)), "IDX");
+
+        publicSchema = createSchema(tbl);
+    }
+
+    /** */
+    @Test
+    public void testIndexRebuild() throws Exception {
+        String sql = "SELECT * FROM TBL WHERE id = 0";
+
+        assertPlan(sql, publicSchema, isInstanceOf(IgniteIndexScan.class));
+
+        tbl.markIndexRebuildInProgress(true);
+
+        assertPlan(sql, publicSchema, isInstanceOf(IgniteTableScan.class));
+
+        tbl.markIndexRebuildInProgress(false);
+
+        assertPlan(sql, publicSchema, isInstanceOf(IgniteIndexScan.class));
+    }
+
+    /** */
+    @Test
+    public void testConcurrentIndexRebuildStateChange() throws Exception {
+        String sql = "SELECT * FROM TBL WHERE id = 0";
+
+        AtomicBoolean stop = new AtomicBoolean();
+
+        IgniteInternalFuture<?> fut = GridTestUtils.runAsync(() -> {
+            while (!stop.get()) {
+                tbl.markIndexRebuildInProgress(true);
+                tbl.markIndexRebuildInProgress(false);
+            }
+        });
+
+        try {
+            for (int i = 0; i < 1000; i++) {
+                IgniteRel rel = physicalPlan(sql, publicSchema);
+
+                assertTrue(rel instanceof IgniteTableScan || rel instanceof IgniteIndexScan);
+            }
+        }
+        finally {
+            stop.set(true);
+        }
+
+        fut.get();
+    }
+
+    /** */
+    @Test
+    public void testIndexScanRewriter() throws Exception {
```

Comment:
I would prefer to move this test to a separate test class (LogicalRelImplementorSelfTest?)

Reply:
It still will be a subclass of AbstractPlannerTest since it requires planning methods. There will be code duplication with the current test (we need IndexRebuildAwareTable and table initialization). Is it ok?

Reply:
If you need a template rel, you could use its serialized form and just deserialize it whenever you need it. As for code duplication, the table class could be moved to a separate class.

Reply:
Fixed